streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/HighAvailabilityTaskAssignor.java

```diff
@@ -184,40 +185,34 @@ public boolean assign() {
-        final Map<TaskId, Integer> unassignedStandbyTasks =
-            configs.numStandbyReplicas == 0 ?
-                Collections.emptyMap() :
-                new HashMap<>(statefulTasksToRankedCandidates.keySet().stream()
-                                  .collect(Collectors.toMap(task -> task, task -> configs.numStandbyReplicas)));
-
-        for (final Map.Entry<UUID, ClientState> clientEntry : clientStates.entrySet()) {
-            final UUID client = clientEntry.getKey();
-            final ClientState state = clientEntry.getValue();
-            final Set<TaskId> prevActiveTasks = state.prevActiveTasks();
-
-            // Verify that this client was caught-up on all stateful active tasks
-            for (final TaskId activeTask : prevActiveTasks) {
-                if (!taskIsCaughtUpOnClient(activeTask, client)) {
-                    return false;
+        final Map<TaskId, Set<UUID>> assignments = new TreeMap<>();
+        for (final TaskId taskId : allTasks) {
+            assignments.put(taskId, new TreeSet<>());
+        }
+        for (final Map.Entry<UUID, ClientState> entry : clientStates.entrySet()) {
+            for (final TaskId activeTask : entry.getValue().prevActiveTasks()) {
+                if (assignments.containsKey(activeTask)) {
+                    assignments.get(activeTask).add(entry.getKey());
                 }
             }
-            unassignedActiveTasks.removeAll(prevActiveTasks);
-
-            if (!unassignedStandbyTasks.isEmpty()) {
-                for (final TaskId task : state.prevStandbyTasks()) {
-                    final Integer remainingStandbys = unassignedStandbyTasks.get(task);
-                    if (remainingStandbys != null) {
-                        if (remainingStandbys == 1) {
-                            unassignedStandbyTasks.remove(task);
-                        } else {
-                            unassignedStandbyTasks.put(task, remainingStandbys - 1);
-                        }
-                    }
-                }
+            for (final TaskId standbyTask : entry.getValue().prevStandbyTasks()) {
+                assignments.get(standbyTask).add(entry.getKey());
             }
         }
-        return unassignedActiveTasks.isEmpty() && unassignedStandbyTasks.isEmpty();
+        final Map<TaskId, Set<UUID>> overOrUnderAssigned =
+            assignments.entrySet()
+                       .stream()
+                       .filter(entry -> {
+                           final TaskId taskId = entry.getKey();
+                           if (statefulTasks.contains(taskId)) {
+                               return entry.getValue().size() != (1 + configs.numStandbyReplicas);
+                           } else {
+                               return entry.getValue().size() != 1;
+                           }
+                       })
+                       .collect(entriesToMap(TreeMap::new));
+
+        return overOrUnderAssigned.isEmpty();
```

Comment:
This fixes a bug uncovered by my "drop node" test in which we would erroneously converge on an assignment that had warmup standbys when it should not.